gcs/upload.py

```diff
@@ -280,6 +282,135 @@ def init_write_object_grpc(cls, db, request_iterator, context):
+                first_message == "write_object_spec"
+            ):  # one shot upload (non-resumable)
+                bucket = db.get_bucket(
+                    request.write_object_spec.resource.bucket, context
+                ).metadata
+                upload = cls.__init_first_write_grpc(request, bucket, context)
+            elif upload is None:
+                return testbench.error.invalid(
+                    "Upload missing a first_message field", context
+                )
+
+            if request.HasField("object_checksums"):
+                # The object checksums may appear only in the first message *or* the last message, but not both
+                if first_message is None and request.finish_write == False:
+                    return testbench.error.invalid(
+                        "Object checksums can be included only in the first or last message",
+                        context,
+                    )
+                if object_checksums is not None:
+                    return testbench.error.invalid(
+                        "Duplicate object checksums in upload",
+                        context,
+                    )
+                object_checksums = request.object_checksums
+
+            data = request.WhichOneof("data")
+            if data == "checksummed_data":
+                checksummed_data = request.checksummed_data
+            elif data is None and request.finish_write:
+                # Handles final message with no data to insert.
+                upload.complete = True
+                continue
+
+            content = checksummed_data.content
+            crc32c_hash = (
+                checksummed_data.crc32c if checksummed_data.HasField("crc32c") else None
+            )
+            if crc32c_hash is not None:
+                actual_crc32c = crc32c.crc32c(content)
+                if actual_crc32c != crc32c_hash:
+                    return testbench.error.mismatch(
+                        "crc32c in checksummed data",
+                        crc32c_hash,
+                        actual_crc32c,
+                        context,
+                    )
+
+            # The testbench should ignore any request bytes that have already been persisted,
+            # thus we validate write_offset against persisted_size.
+            # https://github.com/googleapis/googleapis/blob/15b48f9ed0ae8b034e753c6895eb045f436e257c/google/storage/v2/storage.proto#L320-L329
+            if request.flush or should_checkpoint_session:
+                if request.write_offset < len(upload.media):
+                    range_start = len(upload.media) - request.write_offset
+                    content = testbench.common.partial_media(
+                        content, range_end=len(content), range_start=range_start
+                    )
+                upload.media += content
+            if request.finish_write:
+                upload.complete = True
+            if request.state_lookup:
```

Comment:
I think if `request.finish_write == True and request.state_lookup == True` you do not want to yield just the persisted bytes?

Reply:
Good catch. I've updated this part of the logic so a response with persisted_size is only yielded for **incomplete uploads**. Definitely not elegant... my thought process here is for completed uploads when `finish_write = True` to be finalized out of the loop (store full object checksums, create new object, insert to database, and yield final response with resource metadata)... thoughts?

Reply:
SGTM